source/modes/window.c

```diff
@@ -1062,21 +1063,24 @@ static cairo_surface_t *_get_icon(const Mode *sw, unsigned int selected_line,
     c->icon = x11_helper_get_screenshot_surface_window(c->window, size);
     c->thumbnail_checked = TRUE;
   }
-  if (c->icon == NULL && c->icon_checked == FALSE) {
-    c->icon = get_net_wm_icon(rmpd->ids->array[selected_line], size);
-    c->icon_checked = TRUE;
-  }
-  if (c->icon == NULL && c->class) {
+  if (c->class) {
     if (c->icon_fetch_uid > 0) {
-      return rofi_icon_fetcher_get(c->icon_fetch_uid);
+      icon = rofi_icon_fetcher_get(c->icon_fetch_uid);
+    } else {
+      char *class_lower = g_utf8_strdown(c->class, -1);
+      c->icon_fetch_uid = rofi_icon_fetcher_query(class_lower, size);
+      g_free(class_lower);
+      icon = rofi_icon_fetcher_get(c->icon_fetch_uid);
     }
-    char *class_lower = g_utf8_strdown(c->class, -1);
-    c->icon_fetch_uid = rofi_icon_fetcher_query(class_lower, size);
-    g_free(class_lower);
-    c->icon_fetch_size = size;
-    return rofi_icon_fetcher_get(c->icon_fetch_uid);
   }
   c->icon_fetch_size = size;
+
+  if (icon != NULL && config.window_prefer_icon_theme) {
+    return icon;
+  } else if (c->icon == NULL && c->icon_checked == FALSE) {
+    c->icon = get_net_wm_icon(rmpd->ids->array[selected_line], size);
+    c->icon_checked = TRUE;
+  }
```

Comment:
Is this logic correct?  when `config.window_prefer_icon_theme` is not set, it will never fall back to the 'class icon' when the application has no `net_wm_icon` set.
It now always requests this icon, even when not used. This seems wasteful.

Reply:
Tried to address this in [4bd7d3b](https://github.com/davatorium/rofi/pull/1702/commits/4bd7d3b59d70db4bc8e2f1f01305828c79f45e86)